Source/Methods/Envelope Tracking/EnvelopeTracking.cs

```diff
@@ -0,0 +1,309 @@
+
+            // get lookup table input power trace in 1ohm watts
+            float[] lutDutInputPowerWattOneOhm = new float[lookUpTable.DutInputPower_dBm.Length];
+            for (int i = 0; i < lookUpTable.DutInputPower_dBm.Length; i++)
+                lutDutInputPowerWattOneOhm[i] = (float)Math.Pow(10.0, (lookUpTable.DutInputPower_dBm[i] - 10.0) / 10.0); // V^2 = 10^((Pin - 10.0)/10)
+
+            // run the trace through 1D interpolation
+            float[] rawEnvelope = LinearInterpolation1D(lutDutInputPowerWattOneOhm, lookUpTable.SupplyVoltage_V, powerTrace_W);
+
+            // create waveform to return to the user
+            Waveform envelopeWaveform = new Waveform()
+            {
+                Name = referenceWaveform.Name + "Envelope",
+                Data = referenceWaveform.Data.Clone(),
+                SignalBandwidth_Hz = 0.8 * referenceWaveform.SampleRate,
+                PAPR_dB = double.NaN, // unnecessary to calculate as it won't be used
+                BurstLength_s = double.NaN, // not applicable in ET
+                SampleRate = referenceWaveform.SampleRate,
+                // burst start and stop locations will be left as null since they also will not be used
+                IdleDurationPresent = referenceWaveform.IdleDurationPresent, // can't null a bool, so copy from reference waveform
+                RuntimeScaling = 10 * Math.Log10(0.9), // applies 10% headroom to the waveform at runtime
+                Script = referenceWaveform.Script // we will copy the script but call replace on it in the following lines to change the waveform name
+            };
+            envelopeWaveform.Script = envelopeWaveform.Script.Replace(referenceWaveform.Name, envelopeWaveform.Name); // this is better here since now we only have to add the suffix once
+            WritableBuffer<ComplexSingle> envWfmWriteBuffer = envelopeWaveform.Data.GetWritableBuffer();
+
+            // copy raw envelope data into cloned envelope waveform
+            for (int i = 0; i < rawEnvelope.Length; i++)
+                envWfmWriteBuffer[i] = ComplexSingle.FromSingle(rawEnvelope[i]);
+
+            return envelopeWaveform;
+        }
+        #endregion
+
+        #region Instrument Configuration
+        public static void ConfigureEnvelopeGenerator(NIRfsg envVsg, EnvelopeGeneratorConfiguration envVsgConfig, TrackerConfiguration trackerConfig)
+        {
+            if (envVsgConfig.TerminalConfiguration == RfsgTerminalConfiguration.Differential)
+                envVsg.IQOutPort[""].LoadImpedance = trackerConfig.InputImpedance_Ohms == 50.0 ? 100.0 : trackerConfig.InputImpedance_Ohms;
+            else
+                envVsg.IQOutPort[""].LoadImpedance = trackerConfig.InputImpedance_Ohms;
+            envVsg.IQOutPort[""].TerminalConfiguration = envVsgConfig.TerminalConfiguration;
+            envVsg.IQOutPort[""].CommonModeOffset = trackerConfig.CommonModeOffset_V;

```

Comment:
Based on my understanding [from the documentation](http://zone.ni.com/reference/en-XX/help/371025V-01/cvirfsg/nirfsg_attr_iq_out_port_terminal_configuration/) (sorry could only find the C API), the 5820 only supports differential. 

>PXIe-5820: The only valid value for this attribute is NIRFSG_VAL_DIFFERENTIAL.

If so, this code is entirely unneeded. 

I know that you're getting this from the RFIC examples, because I remember this is configured there. @NISystemsEngineering/rfic-systems-r-d  can you comment on this? I am not sure why this is included in the examples in the first place because it doesn't appear to be supported.

Reply:
Agreed we can remove it, hard set it, and add comment

Reply:
9561a5e4bd08e0f93528dfb8506b35d36dc549d7

Reply:
Sounds good. @mkrmasch and @manderseck agreed that it did not seem correct for it to be there and recommended removing it.

Reply:
9561a5e4bd08e0f93528dfb8506b35d36dc549d7